liminal/actions/messages.ts

```diff
@@ -0,0 +1,99 @@
+import type { Spec } from "../Spec.ts"
+import { fixTemplateStrings } from "../util/fixTemplateStrings.ts"
+import { ActionBase, type ActionEventBase } from "./actions_base.ts"
+import type {
+  FilePart,
+  ImagePart,
+  ReasoningPart,
+  RedactedReasoningPart,
+  TextPart,
+  ToolCallPart,
+} from "./content_part.ts"
+
+export type Message = SystemMessage | UserMessage | AssistantMessage | ToolMessage
+
+// Shared base types
+type MessageType = "assistant_message" | "system_message" | "user_message" | "tool_message"
+type MessageEventType = "assistant_messaged" | "system_messaged" | "user_messaged" | "tool_messaged"
+
+interface BaseMessage<T extends MessageType, C, S extends Spec = Spec> extends ActionBase<T, S> {
+  content: C
+}
+
+interface BaseMessagedEvent<T extends MessageEventType, C> extends ActionEventBase<T> {
+  content: C
+}
+
+const messageFn = <
+  TMessage extends BaseMessage<any, any, { Entry: any; Event: BaseMessagedEvent<any, any> }>,
+>(
+  messageType: TMessage["action"],
+  eventType: TMessage[""]["Event"]["type"],
+) =>
+  function*(
+    ...[raw, ...substitutions]: [content: TMessage["content"]] | [
+      raw: TemplateStringsArray,
+      ...substitutions: Array<string>,
+    ]
+  ): Generator<TMessage, void> {
+    const content = raw instanceof Object && "raw" in raw && Array.isArray(raw.raw)
+      ? String.raw(fixTemplateStrings(raw as TemplateStringsArray), ...substitutions)
+      : raw as TMessage["content"]
+    // @ts-ignore, you can't please everyone
+    return yield ActionBase(messageType, {
```

Comment:
suspicious, but I hope it is acceptable to now have the ability to consolidate all the functions into one simple implementation

Reply:
It's acceptable. Consolidating into a single `messages` file makes a lot of sense.